distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Hidden_Table_Registry.enso

```diff
@@ -0,0 +1,71 @@
+from Standard.Base import all
+import Standard.Base.Runtime.Managed_Resource.Managed_Resource
+import Standard.Base.Runtime.Ref.Ref
+
+## PRIVATE
+   A reference to a hidden table that keeps it alive.
+
+   Once all references to a particular hidden table are garbage collected, the
+   hidden table itself will be dropped.
+type Hidden_Table_Reference
+    Reference (parent_mapping : Ref (Map Text Integer)) (table_name : Text) (drop : Text -> Nothing)
+
+## PRIVATE
+   A registry that keeps track of temporary hidden tables.
+
+   These tables will all be destroyed once the connection is closed, but to
+   avoid creating too many, the registry tries to drop them more eagerly once
+   they stop being needed.
+
+   ! Concurrency
+
+     Note that this code is run in GC threads and may run into concurrency
+     issues. Currently, we have no synchronization primitives in Enso to prevent
+     this. This is just a 'heuristic' helper that tries to limit the amount of
+     dry-run tables living in a session, so the potential failures should not be
+     fatal. Nonetheless, it may be worth to revise this once we get better
+     concurrency support.
+
+     The potential point of failure is the two places were we `get` and later
+     `put` an updated `Ref`. These can happen concurrently and result in the
+     mapping getting inconsistent.
```

Comment:
Given this is an internal class - does it make sense to do this within polyglot Java. We can then make it atomic and thread safe. 

This isn't user facing and paying the polyglot cost is minimal here I think.

Reply:
Yeah, I think we could do that.